coil-core/src/androidMain/kotlin/coil3/decode/FastImageDecoderDecoder.kt

```diff
@@ -0,0 +1,194 @@
+                            dstWidth = dstWidth,
+                            dstHeight = dstHeight,
+                            scale = options.scale,
+                        )
+
+                        // Set the target size if the image is larger than the requested dimensions
+                        // or the request requires exact dimensions.
+                        isSampled = multiplier < 1
+                        if (isSampled || !options.allowInexactSize) {
+                            val targetWidth = (multiplier * srcWidth).roundToInt()
+                            val targetHeight = (multiplier * srcHeight).roundToInt()
+                            setTargetSize(targetWidth, targetHeight)
+                        }
+                    }
+
+                    // Configure any other attributes.
+                    configureImageDecoderProperties()
+                }
+            } finally {
+                imageDecoder?.close()
+                toClose.close()
+            }
+        }
+        return DecodeResult(
+            image = image,
+            isSampled = isSampled,
+        )
+    }
+
+    private fun ImageDecoder.configureImageDecoderProperties() {
+        // https://github.com/element-hq/element-android/pull/7184
+        val ignoreHardware = Build.VERSION.SDK_INT < Build.VERSION_CODES.Q
+        allocator = if (options.bitmapConfig.isHardware && !ignoreHardware) {
+            ImageDecoder.ALLOCATOR_HARDWARE
+        } else {
+            ImageDecoder.ALLOCATOR_SOFTWARE
+        }
+        memorySizePolicy = if (options.allowRgb565) {
+            ImageDecoder.MEMORY_POLICY_LOW_RAM
+        } else {
+            ImageDecoder.MEMORY_POLICY_DEFAULT
+        }
+        if (options.colorSpace != null) {
+            setTargetColorSpace(options.colorSpace)
+        }
+        isUnpremultipliedRequired = !options.premultipliedAlpha
+        postProcessor = options.animatedTransformation?.asPostProcessor()
+    }
+
+    private suspend fun wrapDrawable(baseDrawable: Drawable): Drawable {
+        if (baseDrawable !is AnimatedImageDrawable) {
+            return baseDrawable
+        }
+
+        baseDrawable.repeatCount = options.repeatCount
+
+        // Set the start and end animation callbacks if any one is supplied through the request.
+        val onStart = options.animationStartCallback
+        val onEnd = options.animationEndCallback
+        if (onStart != null || onEnd != null) {
+            // Animation callbacks must be set on the main thread.
+            withContext(Dispatchers.Main.immediate) {
+                baseDrawable.registerAnimationCallback(animatable2CallbackOf(onStart, onEnd))
+            }
+        }
+
+        // Wrap AnimatedImageDrawable in a ScaleDrawable so it always scales to fill its bounds.
+        return ScaleDrawable(baseDrawable, options.scale)
+    }
+}
+
+@RequiresApi(28)
+class FastImageDecoderFactory @JvmOverloads constructor(
```

Comment:
This should be a nested class to be consistent with other decoders: `FastImageDecoder.Factory`.

Reply:
Done